common/src/test/java/org/apache/drill/test/BaseTest.java

```diff
@@ -15,24 +15,25 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.drill.exec.expr;
+package org.apache.drill.test;
 
-import org.apache.drill.BaseTestQuery;
-import org.apache.drill.common.util.TestTools;
-import org.junit.Test;
+import org.apache.drill.common.util.GuavaPatcher;
+import org.apache.drill.common.util.ProtobufPatcher;
 
-public class TestPrune extends BaseTestQuery {
+public class BaseTest {
 
-  String MULTILEVEL = TestTools.getWorkingPath() + "/../java-exec/src/test/resources/multilevel";
-
-  @Test
-  public void pruneCompound() throws Exception {
-    test(String.format("select * from dfs.`%s/csv` where x is null and dir1 in ('Q1', 'Q2')", MULTILEVEL));
-  }
-
-  @Test
-  public void pruneSimple() throws Exception {
-    test(String.format("select * from dfs.`%s/csv` where dir1 in ('Q1', 'Q2')", MULTILEVEL));
+  static {
+    /*
+     * HBase and MapR-DB clients use older version of protobuf,
+     * and override some methods that became final in recent versions.
+     * This code removes these final modifiers.
```

Comment:
Shouldn't these clients be addressed directly? We're doing patching for tests, but what about production code? Possible solutions:

* Load these clients into their own class loader so their dependencies do not conflict.
* Shade them.
* Upgrade to newer versions that use more recent Guava libraries.
* Do the patching within Drill: perhaps have a plugin-specific startup event that can trigger patching where needed.
* Lobby the creators of the libraries to fix the problems.

My concern is that changing all our tests to work around bugs in external libraries is an unstable solution: we'll find ourselves doing more of this each time some new external conflict arises.

Reply:
We're doing patching both in tests and in production (see Drillbit class). 
This is actually quite extensive topic. We already have shaded Guava that is used by Drill. The problem is: some of libraries we're using have transitive dependencies to different Guava versions. In total we have 3 or 4 versions we should make friends with. So shading is not the way. I guess we can do some playing with classloaders, but it'll be tricky and additionally we'll have bloated jars.
As for your last paragraph, this would definitely be great if every library had the same dependencies versions or if Guava was fully backward compatible... Sadly, there is nothing we can do on this account.

Reply:
Good ideas, but I have some questions:
1. I'm not sure how to tell maven to keep a specific version to use it in the future but do not use it for other libraries.
2. When I was shading Guava for Drill usage, I considered this approach, but we have a lot of libraries with their specific versions, for example, hbase, hadoop-common, hadoop-client, hadoop-hdfs. It would be redundant to shade them all.
3. Unfortunately, even the latest versions of most libraries use obsolete guava versions (I have checked hbase).